[14.0][IMP] Backport of 16.0 Improvements & Adjusted Tests

raumschmiede-joshuaL opened the pull request:
Hello.

I split up my other PR https://github.com/OCA/server-tools/pull/3346 to add the new field here.

And I backported a few of your commits that improve the module, @lmignon. 

EDIT: Now it is mostly just a backport. My other commits improve the tests a bit.

@mt-software-de

lmignon committed:
[IMP] base_exception: One model per python file

lmignon committed:
[IMP] base_exception: Improves performances

Before this change, the inverse relation from the exception to the records linked to the exception was read to determine if the rule should be added or removed from the rule detected for the current records. On large database with non blocking exceptions this lead to a performance issues since this issues a read of all the SO where the exception has been applied. With this change, we only read the information from the current records

lmignon committed:
[IMP] base_exception: Improves performances

Keeps rule information into cache to avoid to always call the ORM for static informations. This is change allows a boost into the evaluation of exception rules by avoiding useless SQL queries to retrieve rules definitions

lmignon committed:
[IMP] base_exception: Improves performances

Don't call the ORM to filter current recordset. Replace call to the orm by a call to the filtered_domain method.

raumschmiede-joshuaL committed:
[IMP] base_exception: Moved test setup into common

raumschmiede-joshuaL committed:
[IMP] base_exception: Improved tests

raumschmiede-joshuaL commented on base_exception/models/base_exception_method.py:
```diff
@@ -0,0 +1,158 @@
+        for rule_info in rules_info:
+            main_records = self._get_main_records()
+            records_with_rule_in_exceptions = main_records.filtered(
+                lambda r, rule_id=rule_info.id: rule_id in r.exception_ids.ids
```
@lmignon, I saw this in your change. Why `rule_id` as optional param and not like this?

```suggestion
                lambda r: rule_info.id in r.exception_ids.ids
```

hparfr replied:
https://docs.python-guide.org/writing/gotchas/#late-binding-closures

raumschmiede-joshuaL commented:
FYI @rousseldenis, because you reviewed PR https://github.com/OCA/server-tools/pull/2670 that introduced the improvements

mt-software-de commented:
Thx. Please also separate this PR into 2. One for the Backport one for your Feature request.

raumschmiede-joshuaL commented:
@mt-software-de, I will create another PR when this here was clarified

lmignon reviewed (approved):
LGTM, code review only

raumschmiede-joshuaL commented:
Sorry, I changed something again. I moved the new field into the other model. Makes more sense and now the unit tests of `sale_exception` work without any adjustments

raumschmiede-joshuaL commented:
@lmignon and @hparfr, I removed the commit with the new field for now. I want to improve the module to have a better way to handle sub-exceptions.
The new field is probably still needed but somewhere else in the code. I will create a new PR for it later.

But this here can be merged. I will try to not change something again here :D

mt-software-de reviewed (approved):
LGTM.
Could you also create FWP PRs for the newer versions of your refactor commits?

raumschmiede-joshuaL commented:
@hparfr and @lmignon, can you check this here again? Then we can merge it

hparfr commented:
/ocabot merge major
